src/spaceTruckerApplication.js

```diff
@@ -0,0 +1,129 @@
+
+import AppStates from "./appstates"
+import logger from "./logger"
+import MainMenuScene from "./mainMenuScene";
+
+
+class SpaceTruckerApplication {
+    *appStateMachine() {
+        let previousState = null;
+        let currentState = null;
+        function setState(newState) {
+            previousState = currentState;
+            currentState = newState;
+            logger.logInfo("App state changed. Previous state:" + previousState + " New state: " + newState);
+            return newState;
+        }
+
+        while (true) {
+            let nextState = yield;
+            if (nextState !== null && nextState !== undefined) {
+                setState(nextState);
+                if (nextState === AppStates.EXITING) {
+                    return currentState;
+                }
+            }
+        }
+    }
+
+    get currentState() {
+        return this._stateMachine.next();
+    }
+
+    get activeScene() {
+        return this._currentScene;
+    }
+
+
+    moveNextAppState(state) {
+        return this._stateMachine.next(state).value;
+    }
+
+    constructor(engine) {
+        this._engine = engine;
+        this._currentScene = null;
+        this._stateMachine = this.appStateMachine();
+        this._mainMenu = null;
+
+        this.moveNextAppState(AppStates.CREATED);
+    }
+
+    async initialize() {
+        this._engine.enterFullscreen(true);
+
+        // note: this will be replaced with the call done internally from AssetManager at some point
+        this._engine.displayLoadingUI();
+        
+        this.moveNextAppState(AppStates.INITIALIZING)
+
+        // for simulating loading times
+        const p = new Promise((res, rej) => {
+            setTimeout(() => res(), 5000);
+        });
+
+        await p;
+
+        this._engine.hideLoadingUI();
+
+       
+    }
+
+    async run() {
+        await this.initialize();
+        await this.goToMainMenu();
+
+        this._engine.runRenderLoop(() => {
```

Comment:
@RaananW -- I realized looking over this that the `run` method returns (almost) immediately after the call to `this._engine.runRenderLoop`, meaning that essentially the entirety of the application's state is being captured into the arrow fn. 

Do you see this being a potential issue? Should the run method "block" until the render loop exits instead?

Reply:
I don't see any issue at all. The render loop is the important thing here and not its first execution. It should never block and should execute as fast as it can.